constants/constants.go

```diff
@@ -51,14 +27,46 @@ var (
 	// MagicDirectives are directives automatically appended to Dockerfiles
 	// when pushing the image. These directives allow the built image to be
 	// 're-used'.
-	MagicDirectives = fmt.Sprintf(`
-COPY --chmod=0755 %[1]s %[2]s
-COPY --chmod=0644 %[3]s %[4]s
+	MagicDirectives = `
+COPY --chmod=0755 .envbuilder.tmp/envbuilder /.envbuilder/bin/envbuilder
+COPY --chmod=0644 .envbuilder.tmp/image /.envbuilder/image
 USER root
 WORKDIR /
-ENTRYPOINT [%[2]q]
-`,
-		".envbuilder.tmp/envbuilder", MagicBinaryLocation,
-		".envbuilder.tmp/image", MagicImage,
-	)
+ENTRYPOINT ["/.envbuilder/bin/envbuilder"]
+`
 )
+
+// ErrNoFallbackImage is returned when no fallback image has been specified.
+var ErrNoFallbackImage = errors.New("no fallback image has been specified")
+
+// MagicDir is a working directory for envbuilder. We use this to
+// store files that are used when building images.
+type MagicDir string
+
+// String returns the string representation of the MagicDir.
+func (m MagicDir) String() string {
+	if m == "" {
+		// Instead of the zero value, use defaultMagicDir.
+		return defaultMagicDir
+	}
+	return filepath.Join("/", string(m))
+}
```

Comment:
I'm thinking this abstraction could be a source of bugs. One might not always think to call `String` and instead do `string(MagicDir)` which would result in an empty string or potentially lack of `/`.

I'd rather see this as a validation step in options so that the input is verified there, and an error can be returned if it's not absolute. The options struct could have the `MagicImage` and `MagicBuilt` functions if that's preferable.

Reply:
I think a simpler option here it to simply change the type to a struct with a single unexported field. I'm less worried about leaking the implementation details; we can simply move `MagicDir` to an internal package.